config.txt
{
    "path_checkpoints": [
        "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/checkpoints"
    ],
    "path_loras": [
        "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/loras"
    ],
    "path_embeddings": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/embeddings",
    "path_vae_approx": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/vae_approx",
    "path_vae": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/vae",
    "path_upscale_models": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/upscale_models",
    "path_inpaint": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/inpaint",
    "path_controlnet": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/controlnet",
    "path_clip_vision": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/clip_vision",
    "path_fooocus_expansion": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/prompt_expansion/ai_image_generator_expansion",
    "path_wildcards": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/wildcards",
    "path_safety_checker": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/safety_checker",
    "path_sam": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/sam",
    "path_outputs": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/outputs"
}

config_modification_tutorial.txt
You can modify your "/Users/yukthesh/Downloads/Infinity∞Projects/Image-generator/config.txt" using the below keys, formats, and examples.
Do not modify this file. Modifications in this file will not take effect.
This file is a tutorial and example. Please edit "/Users/yukthesh/Downloads/Infinity∞Projects/Image-generator/config.txt" to really change any settings.
Remember to split the paths with "\\" rather than "\", and there is no "," before the last "}". 


{
    "path_checkpoints": [
        "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/checkpoints"
    ],
    "path_loras": [
        "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/loras"
    ],
    "path_embeddings": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/embeddings",
    "path_vae_approx": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/vae_approx",
    "path_vae": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/vae",
    "path_upscale_models": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/upscale_models",
    "path_inpaint": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/inpaint",
    "path_controlnet": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/controlnet",
    "path_clip_vision": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/clip_vision",
    "path_fooocus_expansion": "/Users/yukthesh/Downloads/Infinity∞Projects/Image-generator/models/prompt_expansion/ai_image_generator_expansion",
    "path_wildcards": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/wildcards",
    "path_safety_checker": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/safety_checker",
    "path_sam": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/models/sam",
    "path_outputs": "/Users/yukthesh/Downloads/Infinity\u221eProjects/Image-generator/outputs",
    "temp_path": "/var/folders/9_/bg1j9c5j30d0p0znx5xrcsy00000gn/T/ai_image_generator",
    "temp_path_cleanup_on_launch": true,
    "default_model": "juggernautXL_v8Rundiffusion.safetensors",
    "previous_default_models": [
        "juggernautXL_version8Rundiffusion.safetensors",
        "juggernautXL_version7Rundiffusion.safetensors",
        "juggernautXL_v7Rundiffusion.safetensors",
        "juggernautXL_version6Rundiffusion.safetensors",
        "juggernautXL_v6Rundiffusion.safetensors"
    ],
    "default_refiner": "None",
    "default_refiner_switch": 0.5,
    "default_loras_min_weight": -2,
    "default_loras_max_weight": 2,
    "default_loras": [
        [
            true,
            "sd_xl_offset_example-lora_1.0.safetensors",
            0.1
        ],
        [
            true,
            "None",
            1.0
        ],
        [
            true,
            "None",
            1.0
        ],
        [
            true,
            "None",
            1.0
        ],
        [
            true,
            "None",
            1.0
        ]
    ],
    "default_max_lora_number": 5,
    "default_cfg_scale": 4.0,
    "default_sample_sharpness": 2.0,
    "default_sampler": "dpmpp_2m_sde_gpu",
    "default_scheduler": "karras",
    "default_vae": "Default (model)",
    "default_styles": [
        "AI Image Generator V2",
        "AI Image Generator Enhance",
        "AI Image Generator Sharp"
    ],
    "default_prompt_negative": "",
    "default_prompt": "",
    "default_performance": "Speed",
    "default_image_prompt_checkbox": false,
    "default_enhance_checkbox": false,
    "default_advanced_checkbox": false,
    "default_developer_debug_mode_checkbox": false,
    "default_image_prompt_advanced_checkbox": false,
    "default_max_image_number": 32,
    "default_output_format": "png",
    "default_image_number": 2,
    "checkpoint_downloads": {
        "juggernautXL_v8Rundiffusion.safetensors": "https://huggingface.co/lllyasviel/fav_models/resolve/main/fav/juggernautXL_v8Rundiffusion.safetensors"
    },
    "lora_downloads": {
        "sd_xl_offset_example-lora_1.0.safetensors": "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/resolve/main/sd_xl_offset_example-lora_1.0.safetensors"
    },
    "embeddings_downloads": {},
    "vae_downloads": {},
    "available_aspect_ratios": [
        "704*1408",
        "704*1344",
        "768*1344",
        "768*1280",
        "832*1216",
        "832*1152",
        "896*1152",
        "896*1088",
        "960*1088",
        "960*1024",
        "1024*1024",
        "1024*960",
        "1088*960",
        "1088*896",
        "1152*896",
        "1152*832",
        "1216*832",
        "1280*768",
        "1344*768",
        "1344*704",
        "1408*704",
        "1472*704",
        "1536*640",
        "1600*640",
        "1664*576",
        "1728*576"
    ],
    "default_aspect_ratio": "1152*896",
    "default_inpaint_engine_version": "v2.6",
    "default_selected_image_input_tab_id": "uov_tab",
    "default_uov_method": "Disabled",
    "default_controlnet_image_count": 4,
    "default_ip_image_1": "None",
    "default_ip_type_1": "ImagePrompt",
    "default_ip_stop_at_1": 0.5,
    "default_ip_weight_1": 0.6,
    "default_ip_image_2": "None",
    "default_ip_type_2": "ImagePrompt",
    "default_ip_stop_at_2": 0.5,
    "default_ip_weight_2": 0.6,
    "default_ip_image_3": "None",
    "default_ip_type_3": "ImagePrompt",
    "default_ip_stop_at_3": 0.5,
    "default_ip_weight_3": 0.6,
    "default_ip_image_4": "None",
    "default_ip_type_4": "ImagePrompt",
    "default_ip_stop_at_4": 0.5,
    "default_ip_weight_4": 0.6,
    "default_inpaint_advanced_masking_checkbox": false,
    "default_inpaint_method": "Inpaint or Outpaint (default)",
    "default_cfg_tsnr": 7.0,
    "default_clip_skip": 2,
    "default_overwrite_step": -1,
    "default_overwrite_switch": -1,
    "default_overwrite_upscale": -1,
    "example_inpaint_prompts": [
        "highly detailed face",
        "detailed girl face",
        "detailed man face",
        "detailed hand",
        "beautiful eyes"
    ],
    "example_enhance_detection_prompts": [
        "face",
        "eye",
        "mouth",
        "hair",
        "hand",
        "body"
    ],
    "default_enhance_tabs": 3,
    "default_enhance_uov_method": "Disabled",
    "default_enhance_uov_processing_order": "Before First Enhancement",
    "default_enhance_uov_prompt_type": "Original Prompts",
    "default_sam_max_detections": 0,
    "default_black_out_nsfw": false,
    "default_save_only_final_enhanced_image": false,
    "default_save_metadata_to_images": false,
    "default_metadata_scheme": "ai_image_generator",
    "metadata_created_by": "",
    "default_invert_mask_checkbox": false,
    "default_inpaint_mask_model": "isnet-general-use",
    "default_enhance_inpaint_mask_model": "sam",
    "default_inpaint_mask_cloth_category": "full",
    "default_inpaint_mask_sam_model": "vit_b",
    "default_describe_apply_prompts_checkbox": true,
    "default_describe_content_type": [
        "Photograph"
    ]
}

modules/flags.py
from enum import IntEnum, Enum

disabled = 'Disabled'
enabled = 'Enabled'
subtle_variation = 'Vary (Subtle)'
strong_variation = 'Vary (Strong)'
upscale_15 = 'Upscale (1.5x)'
upscale_2 = 'Upscale (2x)'
upscale_fast = 'Upscale (Fast 2x)'

uov_list = [disabled, subtle_variation, strong_variation, upscale_15, upscale_2, upscale_fast]

enhancement_uov_before = "Before First Enhancement"
enhancement_uov_after = "After Last Enhancement"
enhancement_uov_processing_order = [enhancement_uov_before, enhancement_uov_after]

enhancement_uov_prompt_type_original = 'Original Prompts'
enhancement_uov_prompt_type_last_filled = 'Last Filled Enhancement Prompts'
enhancement_uov_prompt_types = [enhancement_uov_prompt_type_original, enhancement_uov_prompt_type_last_filled]

CIVITAI_NO_KARRAS = ["euler", "euler_ancestral", "heun", "dpm_fast", "dpm_adaptive", "ddim", "uni_pc"]

# ai_image_generator: a1111 (Civitai)
KSAMPLER = {
    "euler": "Euler",
    "euler_ancestral": "Euler a",
    "heun": "Heun",
    "heunpp2": "",
    "dpm_2": "DPM2",
    "dpm_2_ancestral": "DPM2 a",
    "lms": "LMS",
    "dpm_fast": "DPM fast",
    "dpm_adaptive": "DPM adaptive",
    "dpmpp_2s_ancestral": "DPM++ 2S a",
    "dpmpp_sde": "DPM++ SDE",
    "dpmpp_sde_gpu": "DPM++ SDE",
    "dpmpp_2m": "DPM++ 2M",
    "dpmpp_2m_sde": "DPM++ 2M SDE",
    "dpmpp_2m_sde_gpu": "DPM++ 2M SDE",
    "dpmpp_3m_sde": "",
    "dpmpp_3m_sde_gpu": "",
    "ddpm": "",
    "lcm": "LCM",
    "tcd": "TCD",
    "restart": "Restart"
}

SAMPLER_EXTRA = {
    "ddim": "DDIM",
    "uni_pc": "UniPC",
    "uni_pc_bh2": ""
}

SAMPLERS = KSAMPLER | SAMPLER_EXTRA

KSAMPLER_NAMES = list(KSAMPLER.keys())

SCHEDULER_NAMES = ["normal", "karras", "exponential", "sgm_uniform", "simple", "ddim_uniform", "lcm", "turbo", "align_your_steps", "tcd", "edm_playground_v2.5"]
SAMPLER_NAMES = KSAMPLER_NAMES + list(SAMPLER_EXTRA.keys())

sampler_list = SAMPLER_NAMES
scheduler_list = SCHEDULER_NAMES

clip_skip_max = 12

default_vae = 'Default (model)'

refiner_swap_method = 'joint'

default_input_image_tab = 'uov_tab'
input_image_tab_ids = ['uov_tab', 'ip_tab', 'inpaint_tab', 'describe_tab', 'enhance_tab', 'metadata_tab']

cn_ip = "ImagePrompt"
cn_ip_face = "FaceSwap"
cn_canny = "PyraCanny"
cn_cpds = "CPDS"

ip_list = [cn_ip, cn_canny, cn_cpds, cn_ip_face]
default_ip = cn_ip

default_parameters = {
    cn_ip: (0.5, 0.6), cn_ip_face: (0.9, 0.75), cn_canny: (0.5, 1.0), cn_cpds: (0.5, 1.0)
}  # stop, weight

output_formats = ['png', 'jpeg', 'webp']

inpaint_mask_models = ['u2net', 'u2netp', 'u2net_human_seg', 'u2net_cloth_seg', 'silueta', 'isnet-general-use', 'isnet-anime', 'sam']
inpaint_mask_cloth_category = ['full', 'upper', 'lower']
inpaint_mask_sam_model = ['vit_b', 'vit_l', 'vit_h']

inpaint_engine_versions = ['None', 'v1', 'v2.5', 'v2.6']
inpaint_option_default = 'Inpaint or Outpaint (default)'
inpaint_option_detail = 'Improve Detail (face, hand, eyes, etc.)'
inpaint_option_modify = 'Modify Content (add objects, change background, etc.)'
inpaint_options = [inpaint_option_default, inpaint_option_detail, inpaint_option_modify]

describe_type_photo = 'Photograph'
describe_type_anime = 'Art/Anime'
describe_types = [describe_type_photo, describe_type_anime]

sdxl_aspect_ratios = [
    '704*1408', '704*1344', '768*1344', '768*1280', '832*1216', '832*1152',
    '896*1152', '896*1088', '960*1088', '960*1024', '1024*1024', '1024*960',
    '1088*960', '1088*896', '1152*896', '1152*832', '1216*832', '1280*768',
    '1344*768', '1344*704', '1408*704', '1472*704', '1536*640', '1600*640',
    '1664*576', '1728*576'
]


class MetadataScheme(Enum):
    A1111 = 'a1111'
    AI_IMAGE_GENERATOR = 'ai_image_generator'


metadata_scheme = [
    (f'{MetadataScheme.A1111.value} (plain text)', MetadataScheme.A1111.value),
    (f'{MetadataScheme.AI_IMAGE_GENERATOR.value} (json)', MetadataScheme.AI_IMAGE_GENERATOR.value),
]


class OutputFormat(Enum):
    PNG = 'png'
    JPEG = 'jpeg'
    WEBP = 'webp'

    @classmethod
    def list(cls) -> list:
        return list(map(lambda c: c.value, cls))


class PerformanceLoRA(Enum):
    QUALITY = None
    SPEED = None
    EXTREME_SPEED = 'sdxl_lcm_lora.safetensors'
    LIGHTNING = 'sdxl_lightning_4step_lora.safetensors'
    HYPER_SD = 'sdxl_hyper_sd_4step_lora.safetensors'


class Steps(IntEnum):
    QUALITY = 60
    SPEED = 30
    EXTREME_SPEED = 8
    LIGHTNING = 4
    HYPER_SD = 4

    @classmethod
    def keys(cls) -> list:
        return list(map(lambda c: c, Steps.__members__))


class StepsUOV(IntEnum):
    QUALITY = 36
    SPEED = 18
    EXTREME_SPEED = 8
    LIGHTNING = 4
    HYPER_SD = 4


class Performance(Enum):
    QUALITY = 'Quality'
    SPEED = 'Speed'
    EXTREME_SPEED = 'Extreme Speed'
    LIGHTNING = 'Lightning'
    HYPER_SD = 'Hyper-SD'

    @classmethod
    def list(cls) -> list:
        return list(map(lambda c: (c.name, c.value), cls))

    @classmethod
    def values(cls) -> list:
        return list(map(lambda c: c.value, cls))

    @classmethod
    def by_steps(cls, steps: int | str):
        return cls[Steps(int(steps)).name]

    @classmethod
    def has_restricted_features(cls, x) -> bool:
        if isinstance(x, Performance):
            x = x.value
        return x in [cls.EXTREME_SPEED.value, cls.LIGHTNING.value, cls.HYPER_SD.value]

    def steps(self) -> int | None:
        return Steps[self.name].value if self.name in Steps.__members__ else None

    def steps_uov(self) -> int | None:
        return StepsUOV[self.name].value if self.name in StepsUOV.__members__ else None

    def lora_filename(self) -> str | None:
        return PerformanceLoRA[self.name].value if self.name in PerformanceLoRA.__members__ else None
